### PR TITLE
Dynamic pre-provisioning

### PR DESCRIPTION
**Pull request by freva**
* Makes it possible to allocate nodes to hosts in state `provisioned`, `ready` (and `active` as before). 
* Adds a `preprovision-capacity` flag which defines resources and amount that will be kept preprovisioned.

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                .sorted(NodeResourceComparator.memoryDiskCpuOrder())
+                .collect(Collectors.toList());
+
+        int hostI = 0;
+        int capacityI = 0;
+        while (hostI < emptyHosts.size() && capacityI < preProvisionCapacity.size()) {
+            if (NodePrioritizer.ALLOCATABLE_HOST_STATES.contains(emptyHosts.get(hostI).state()) &&
+                    emptyHosts.get(hostI).flavor().resources().satisfies(preProvisionCapacity.get(capacityI))) {
+                emptyHosts.remove(hostI);
+                preProvisionCapacity.remove(capacityI);
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Note: List typically have a rather poor remove(0) performance

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                    return IntStream.range(0, cap.getCount()).mapToObj(i -> resources);
+                })
+                .sorted(NodeResourceComparator.memoryDiskCpuOrder())
+                .collect(Collectors.toList());
+
+        int hostI = 0;
+        int capacityI = 0;
+        while (hostI < emptyHosts.size() && capacityI < preProvisionCapacity.size()) {
+            if (NodePrioritizer.ALLOCATABLE_HOST_STATES.contains(emptyHosts.get(hostI).state()) &&
+                    emptyHosts.get(hostI).flavor().resources().satisfies(preProvisionCapacity.get(capacityI))) {
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: If there is only one empty host HB with capacity satisfying CB from the preprovision-capacity flag (with count 1), and there is also a capacity CA > CB in the flag (with count 1), then this algorithm would allocate CA and CB and deallocate HB. It is unnecessary to deallocate HB and allocate CB.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            try {
+                List<Node> updatedNodes = hostProvisioner.provision(host, children);
+                nodeRepository().write(updatedNodes, lock);
+            } catch (IllegalArgumentException | IllegalStateException e) {
+                log.log(Level.INFO, "Failed to provision " + host.hostname() + ": " + Exceptions.toMessageString(e));
+            } catch (FatalProvisioningException e) {
+                log.log(Level.SEVERE, "Failed to provision " + host.hostname() + ", failing out the host recursively", e);
+                // Fail out as operator to force a quick redeployment
+                nodeRepository().failRecursively(
+                        host.hostname(), Agent.operator, "Failed by HostProvisioner due to provisioning failure");
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: See comment above for why it is `operator`.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java (hunk 1 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Yes, but these will typically be tiny lists, probably less than 5 elements.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java (hunk 2 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Fixed by always looping through all the host and finding the cheapest one that satisfies the capacity.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                emptyHosts.remove(hostI);
+                preProvisionCapacity.remove(capacityI);
+            } else {
+                // If the host is not in one of the allocatable states or the host does not satisfy any of the
+                // pre-provisioning reservations, keep it in the list so it can be deprovisioned
+                hostI++;
+            }
+        }
+
+        // pre-provisioning is best effort, do one host at the time
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Fixed

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+        int hostI = 0;
+        int capacityI = 0;
+        while (hostI < emptyHosts.size() && capacityI < preProvisionCapacity.size()) {
+            if (NodePrioritizer.ALLOCATABLE_HOST_STATES.contains(emptyHosts.get(hostI).state()) &&
+                    emptyHosts.get(hostI).flavor().resources().satisfies(preProvisionCapacity.get(capacityI))) {
+                emptyHosts.remove(hostI);
+                preProvisionCapacity.remove(capacityI);
+            } else {
+                // If the host is not in one of the allocatable states or the host does not satisfy any of the
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Will ignore hosts in `parked` that do not have `wantToDeprovision` set.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                .collect(Collectors.groupingBy(
+                        node -> provisionedHostsByHostname.get(node.parentHostname().get()),
+                        Collectors.toSet()));
+
+        nodesByProvisionedParent.forEach((host, children) -> {
+            try {
+                List<Node> updatedNodes = hostProvisioner.provision(host, children);
+                nodeRepository().write(updatedNodes, lock);
+            } catch (IllegalArgumentException | IllegalStateException e) {
+                log.log(Level.INFO, "Failed to provision " + host.hostname() + ": " + Exceptions.toMessageString(e));
```
[the remark this answers is not in the document]
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java]: Added number of children to error message.

**Pull request merged**